include/eve/detail/function/simd/x86/swizzle.hpp

```diff
@@ -0,0 +1,258 @@
+//==================================================================================================
+/**
+  EVE - Expressive Vector Engine
+  Copyright 2018-2021 Joel FALCOU
+  Copyright 2018-2021 Jean-Thierry LAPRESTE
+
+  Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+  SPDX-License-Identifier: MIT
+**/
+//==================================================================================================
+#pragma once
+
+#include <eve/detail/abi.hpp>
+#include <eve/detail/category.hpp>
+#include <eve/detail/function/simd/common/swizzle_helpers.hpp>
+#include <eve/traits/as_floating_point.hpp>
+#include <eve/traits/as_integer.hpp>
+#include <eve/forward.hpp>
+#include <eve/pattern.hpp>
+
+namespace eve::detail
+{
+  //================================================================================================
+  // shuffle requires cross-api re-targeting (ie AVX512 128bits calling back SSSE3)
+  // so instead of our classical one-size-fit-all overloads, we need to split it into
+  // API supports overloads
+  //================================================================================================
+
+  //================================================================================================
+  // Unary swizzle - logical on AVX512 ABI, call general case on others
+  //================================================================================================
+  template<typename T, typename N, x86_abi ABI, shuffle_pattern Pattern>
+  EVE_FORCEINLINE auto basic_swizzle_ ( EVE_SUPPORTS(sse2_)
+                                      , logical<wide<T,N,ABI>> const& v, Pattern p
+                                      ) noexcept
+  {
+    if constexpr( current_api >= avx512 ) return to_logical((v.mask())[p]);
+    else                                  return basic_swizzle_(EVE_RETARGET(cpu_),v,p);
```

Comment:
Why retarget cpu? I feel that just shuffle mask and bitcast back

Reply:
which is what the CPU logical version does. I call it to prevent code duplication.